File: objective-c/test/Ice/invoke/BlobjectI.m
// **********************************************************************
//
// Copyright (c) 2003-present ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

#import <objc/Ice.h>
#import <invoke/BlobjectI.h>

@implementation BlobjectI
-(BOOL) ice_invoke:(NSData*)inEncaps outEncaps:(NSMutableData**)outEncaps current:(ICECurrent*)current
{
    id<ICECommunicator> communicator = [current.adapter getCommunicator];
    id<ICEInputStream> inS = [ICEUtil createInputStream:communicator data:inEncaps];
    id<ICEOutputStream> outS = [ICEUtil createOutputStream:communicator];
    [outS startEncapsulation];
    if([current.operation isEqualToString:@"opOneway"])
    {
        return YES;
    }
    else if([current.operation isEqualToString:@"opString"])
    {
        [inS startEncapsulation];
        NSString* s = [inS readString];
        [inS endEncapsulation];
        [outS writeString:s];
        [outS writeString:s];
        [outS endEncapsulation];
        *outEncaps = [outS finished];
        return YES;
    }
    else if([current.operation isEqualToString:@"opException"])
    {
        if([current.ctx objectForKey:@"raise"] != nil)
        {
            @throw [TestInvokeMyException myException];
        }
        TestInvokeMyException* ex = [TestInvokeMyException myException];
        [outS writeException:ex];
        [outS endEncapsulation];
        *outEncaps = [outS finished];
        return NO;
    }
    else if([current.operation isEqualToString:@"shutdown"])
    {
        [outS endEncapsulation];
        *outEncaps = [outS finished];
        [communicator shutdown];
        return YES;
    }
    else if([current.operation isEqualToString:@"ice_isA"])
    {
        [inS startEncapsulation];
        NSString* s = [inS readString];
        [inS endEncapsulation];
        if([s isEqualToString:[TestInvokeMyClass ice_staticId]])
        {
            [outS writeBool:YES];
        }
        else
        {
            [outS writeBool:NO];
        }
        [outS endEncapsulation];
        *outEncaps = [outS finished];
        return YES;
    }
    else
    {
        @throw [ICEOperationNotExistException operationNotExistException:__FILE__
                                              line:__LINE__
                                              id:current.id_
                                              facet:current.facet
                                              operation:current.operation];
    }
}
@end

File: java-compat/test/src/main/java/test/Ice/invoke/BlobjectI.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

package test.Ice.invoke;

import test.Ice.invoke.Test.MyException;

public class BlobjectI extends Ice.Blobject
{
    @Override
    public boolean
    ice_invoke(byte[] inParams, Ice.ByteSeqHolder outParams, Ice.Current current)
        throws Ice.UserException
    {
        Ice.Communicator communicator = current.adapter.getCommunicator();
        Ice.InputStream in = new Ice.InputStream(communicator, inParams);
        in.startEncapsulation();
        Ice.OutputStream out = new Ice.OutputStream(communicator);
        out.startEncapsulation();
        if(current.operation.equals("opOneway"))
        {
            outParams.value = new byte[0];
            return true;
        }
        else if(current.operation.equals("opString"))
        {
            String s = in.readString();
            out.writeString(s);
            out.writeString(s);
            out.endEncapsulation();
            outParams.value = out.finished();
            return true;
        }
        else if(current.operation.equals("opException"))
        {
            if(current.ctx.containsKey("raise"))
            {
                throw new MyException();
            }
            MyException ex = new MyException();
            out.writeException(ex);
            out.endEncapsulation();
            outParams.value = out.finished();
            return false;
        }
        else if(current.operation.equals("shutdown"))
        {
            communicator.shutdown();
            return true;
        }
        else if(current.operation.equals("ice_isA"))
        {
            String s = in.readString();
            if(s.equals("::Test::MyClass"))
            {
                out.writeBool(true);
            }
            else
            {
                out.writeBool(false);
            }
            out.endEncapsulation();
            outParams.value = out.finished();
            return true;
        }
        else
        {
            Ice.OperationNotExistException ex = new Ice.OperationNotExistException();
            ex.id = current.id;
            ex.facet = current.facet;
            ex.operation = current.operation;
            throw ex;
        }
    }
}
